Install a functional host helper on OS X

shtrom opened the pull request:
Do some more pattern substitution so the host helper Just Works(TM) on OS X after installation
* (all platforms) detect where the `pass` executable is in the same way as for `python3`, and rewrite it in the host script
* (Darwin) uncomment the `PATH` in the ENV

Signed-off-by: Olivier Mehani <olivier.mehani@learnosity.com>

tuxor1337 commented:
Are you on Mac? Would you please test this pull request: https://github.com/passff/passff-host/pull/7 If it works for you, hardcoding the pass path into the host app is unnecessary anyway.

shtrom commented:
On Thu 29 Mar 2018 at 06:29:55 +0000, Thomas Vogt wrote:
> Are you on Mac? Would you please test this pull request: https://github.com/passff/passff-host/pull/7 If it works for you, hardcoding the pass path into the host app is unnecessary anyway.

At work, yes, unfortunately.

I can test that fix yeah. Will do on Tuesday.

Note that this PR also update the `PATH` in the `COMMAND_ENV` array, but
I'm not sure if it is strictly necessary.

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

5bentz commented:
Thanks for the PR!

Maybe we could just let the PATH uncommented in COMMAND_ENV.
Something like: (sbin is unnecessary)
`PATH="/usr/local/bin:/usr/bin:/bin"`
I quickly tested it on GNU/Linux Debian, pass & passff work as expected.

On a side note:

In GMRL's [.zshrc line 605](https://github.com/grml/grml-etc-core/blob/master/etc/zsh/zshrc#L605), a **MacPorts** is mentioned and it uses other directories.
```
# do MacPorts setup on darwin
if isdarwin && [[ -d /opt/local ]]; then
    # Note: PATH gets set in /etc/zprofile on Darwin, so this can't go into
    # zshenv.
    PATH="/opt/local/bin:/opt/local/sbin:$PATH"
    MANPATH="/opt/local/share/man:$MANPATH"
fi
```
From [macports website](https://guide.macports.org/chunked/introduction.html):
> MacPorts is an easy to use system for compiling, installing, and managing open source software.

shtrom commented:
On Fri 30 Mar 2018 at 14:01:54 +0000, 5bentz wrote:
> In GMRL's [.zshrc line 605](https://github.com/grml/grml-etc-core/blob/master/etc/zsh/zshrc#L605), a **MacPorts** is mentioned and it uses other directories.
> >From [macports website](https://guide.macports.org/chunked/introduction.html):
> > MacPorts is an easy to use system for compiling, installing, and managing open source software. 

Yeah, I don't think MacPorts is widely in use anymore. Homebrew has
kinda taken it all..

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

tuxor1337 commented:
Using some arbitrary value for PATH for all users doesn't sound like a generic solution. However, I like this PR because we do have some Mac users who are confused by the fact, that the host app doesn't work out of the box.

@shtorm: Would you please adapt the code to the current revision to set the PATH variable and the path to the pass script  for Mac users? Thanks in advance!

shtrom commented:
Happy to adapt. However, #7, which uses `which` to detect where `pass` lives sounds like a safer approach, so I'm tempted to only do the `PATH` management, here.

tuxor1337 commented:
I think, #7 is not going to work with Mac, because the PATH variable is not set at the time the `which` command is executed.

shtrom commented:
I'll try it out. But worse comes to worse, we could define the env before using `which`. `/usr/local`, which HomeBrew uses, is also pretty standard across Unices, so perhaps the easiest would be to include it unconditionally.

shtrom commented:
Ok, rebased on `master` and added a few changes:
* The install script now has a sane default path, that it also rewrites in the host script
** It also autodetects if MacPorts pathes are available, but this may cause issues in corner cases, so we may not want that.
** Nonetheless, this is the path used for auto-detecting `pass` afterwards, so it's easily tweaked on weird systems.
* The shipped `passff.py` doesn't specify a path to `pass` by default, but it is unconditionally replaced with whatever `which pass` found in the install script.

tuxor1337 commented on src/install_host_app.sh:
```diff
@@ -126,16 +140,26 @@ else
-#   whereas on GNU sed or BSD sed backup extension may be omit.
-if [ "$KERNEL_NAME" == 'Darwin' ]; then
+if [ "$IS_BSD" = true ]; then
+  # Use BSD style sed on BSD-ish systems
```
We learned from https://github.com/passff/passff-host/pull/12, that  BSD and Darwin shouldn't be handled together. This is probably not going to work.

shtrom replied:
I was actually surprised to see `-i` working at all for BSDs, but apparently OS X and OpenBSD now support in place replacement. I don't know what the status is for other BSDs like Free or Net.

To be honest, though, I was half considering replacing that `if` with a plain POSIX `sed` use that would work in any situation. Should be easy enough to pull of, and that'll clear compatibility concerns.

tuxor1337 commented on src/install_host_app.sh:
```diff
@@ -11,8 +11,14 @@ HOST_URL="https://github.com/passff/passff-host/releases/download/$VERSION/passf
+
 case "$KERNEL_NAME" in
-  *BSD*)
+  *BSD* | *Darwin*)
```
See below, BSD and Darwin should not be handled together.

tuxor1337 commented on src/install_host_app.sh:
```diff
@@ -11,8 +11,14 @@ HOST_URL="https://github.com/passff/passff-host/releases/download/$VERSION/passf
 MANIFEST_URL="https://github.com/passff/passff-host/releases/download/$VERSION/passff.json"
 KERNEL_NAME=$(uname -s)
 
+PATH="/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
```
It's a terrible idea to unconditionally replace the system's `PATH` with this value. It is definitely going to break the install script for a lot of people.

shtrom replied:
Ah, yeah, fair enought. This one is just a reasonably clean standard path, but using the current path makes sense.

tuxor1337 commented on src/install_host_app.sh:
```diff
@@ -126,16 +140,26 @@ else
+  /usr/bin/sed -i '' "/^COMMAND *=/s@\"pass\"@\"$PASS_PATH\"@" "$HOST_FILE_PATH"
+  # Set the PATH to match this script's
+  /usr/bin/sed -i '' "s@\"PATH\":.*@\"PATH\": \"$PATH\"@" "$HOST_FILE_PATH"
+
```
This is fine, but why don't we do this to set the `PATH` to the value it has when the install script is executed? I don't understand why we would replace it by an arbitrary `PATH` that we chose ourselves?!

shtrom commented:
Ok, I simplified the whole `sed` logic so we don't use OS-dependent extensions. We just define both `sed` scripts in variables, and apply them while streaming the source host script and manifest (either with cURL or from the checkout) to the install location.

The replacements in the host scripts are now: the path the `python3`, the path to `pass`, and the `PATH`.

I tested on Debian Buster, OpenBSD 6.2, and OS X High Sierra. All with the proper replacements made in the installed files.

Note that string replacement in the current 1.0.1 script available online will not work out of the box, due to this change b3ce9e77dda81df4227800ab1e7b3934abae703a, so it might be worth either redeploying it after merging this PR, or making the replacement code more tolerant (but it would be more error prone, so I'm not if favour of this).

shtrom commented:
Note that the `--local` installs are still not immediately functional due to the `_VERSIONHOLDER_`.

tuxor1337 commented:
@shtrom Thanks a lot! I will test this later.

Contrary to what you say, if a user installs according to the README, the `--local` option is functional because the README assumes that the user either installs a version from the release page or successfully runs `make` in the `src` directory and then runs the install script from the `build` directory.

shtrom commented:
On Tue 03 Apr 2018 at 22:30:44 -0700, Thomas Vogt wrote:
> Contrary to what you say, if a user installs according to the README, the `--local` option is functional because the README assumes that the user either installs a version from the release page or successfully runs `make` in the `src` directory and then runs the install script from the `build` directory. 

Well. I bet I should read the README, then (:

I was just installing out of src, which was the source of my side issue,
then.

Excellent!

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.
